staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -490,3 +509,78 @@ func startDBSizeMonitorPerEndpoint(client *clientv3.Client, interval time.Durati
 		cancel()
 	}, nil
 }
+
+// checkVersion checks for deprecated versions of etcd and returns an error if find any
+// currently the versions < 3.4.25, between 3.5.0 and 3.5.7
+func checkVersion(c *clientv3.Client, dialTimeOut time.Duration) (err error) {
+	var wg sync.WaitGroup
+
+	eps := c.Endpoints()
+	errc := make(chan error, len(eps))
+	baseCtx := context.TODO()
+
+	ctx, cancel := context.WithCancel(baseCtx)
+	if dialTimeOut > 0 {
+		cancel()
+		ctx, cancel = context.WithTimeout(ctx, dialTimeout)
+	}
+
+	wg.Add(len(eps))
+	for _, ep := range eps {
+		// if cluster is current, any endpoint gives a recent version
+		go func(e string) {
+			defer wg.Done()
+			resp, rerr := c.Status(ctx, e)
+			if rerr != nil {
+				errc <- rerr
+				return
+			}
+			isDeprecated, rerr := isDeprecatedVersion(resp.Version)
+			if err != nil {
+				errc <- rerr
+				return
+			}
+			if isDeprecated {
+				rerr = WarnDeprecatedCluster
+			}
+			errc <- rerr
+		}(ep)
+	}
+	// wait for success
+	for range eps {
+		if err = <-errc; err != nil {
+			break
+		}
+	}
+	cancel()
+	wg.Wait()
+	return err
+}
+
+// isDeprecatedVersion checks for following versions and return true if conditions are true
```

Comment:
Please be more specific, the check here is more about etcd versions affected by https://github.com/etcd-io/etcd/issues/15220.